src/rust/engine/hashing/src/lib.rs

```diff
@@ -312,6 +316,126 @@ impl<W: Write> Write for WriterHasher<W> {
   }
 }
 
+impl<AW: ?Sized + AsyncWrite + Unpin> AsyncWrite for WriterHasher<&mut AW> {
+  fn poll_write(
+    mut self: Pin<&mut Self>,
+    cx: &mut Context<'_>,
+    buf: &[u8],
+  ) -> Poll<Result<usize, Error>> {
+    let inner = Pin::new(&mut *self.inner);
+    let result = inner.poll_write(cx, buf);
+    if let Poll::Ready(Ok(written)) = result {
+      // Hash the bytes that were successfully written.
+      self.hasher.update(&buf[0..written]);
+      self.byte_count += written;
+    }
+    result
+  }
+
+  fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Error>> {
+    let inner = Pin::new(&mut *self.inner);
+    inner.poll_flush(cx)
+  }
+
+  fn poll_shutdown(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Error>> {
+    let inner = Pin::new(&mut *self.inner);
+    inner.poll_shutdown(cx)
+  }
+}
+
+///
+/// Copy the data from reader and hash the bytes in one pass.
+/// Use hash() to just hash without copying the data anywhere.
+///
+pub fn copy_and_hash<R: ?Sized, W: ?Sized>(reader: &mut R, writer: &mut W) -> io::Result<Digest>
+where
+  R: io::Read,
+  W: io::Write,
+{
+  let mut hasher = WriterHasher::new(writer);
+  let _ = io::copy(reader, &mut hasher)?;
+  Ok(hasher.finish().0)
+}
+
+///
+/// Hashes the data into a Digest.
+///
+pub fn hash<R: ?Sized>(reader: &mut R) -> io::Result<Digest>
+where
+  R: io::Read,
+{
+  copy_and_hash(reader, &mut io::sink())
+}
+
+///
+/// Hashes the data into a Digest.
+///
+pub fn hash_path<P: AsRef<Path> + std::fmt::Debug + Clone>(path: P) -> Result<Digest, String> {
+  copy_and_hash(
+    &mut std::fs::File::open(path.clone()).map_err(|e| format!("Failed to open {path:?}: {e}."))?,
+    &mut io::sink(),
+  )
+  .map_err(|e| format!("Failed to hash {path:?}: {e}."))
+}
+
+///
+/// Copy from reader to writer and return whether the copied data matches expected_digest.
+///
+pub fn verified_copy<R: ?Sized, W: ?Sized>(
+  expected_digest: Digest,
+  data_is_immutable: bool,
+  reader: &mut R,
+  writer: &mut W,
+) -> io::Result<bool>
+where
+  R: io::Read,
+  W: io::Write,
+{
+  if data_is_immutable {
+    // Trust that the data hasn't changed, and only validate its length.
+    let copied = io::copy(reader, writer)?;
+    Ok(copied as usize == expected_digest.size_bytes)
+  } else {
+    Ok(expected_digest == copy_and_hash(reader, writer)?)
+  }
```

Comment:
```suggestion
  }
    copy_and_hash(reader, writer).map(|digest| digest == expected_digest)
  }
```

Reply:
Personally I prefer to reason about code through the "happy path". That means short circuiting `Err`s and focusing on the `Ok` paths.